index.rst

```diff
@@ -1,29 +1,375 @@
+If no quota is configured, the quota for all users is unlimited and requests won't be tracked.
+
+Quota overrides
+---------------
+
+Emergency override information will be stored in the Gafaelfawr Redis under the key ``quota-override``.
+The value of the key will be a JSON document such as the following:
+
+.. code-block:: json
+
+   {
+       "default": {
+           "notebook": {
+               "spawn": false,
+               "cpu": 4
+           },
+           "api": {
+               "datalinker": 10
+           }
+       },
+       "groups": {
+           "g_users": {
+               "api": {
+                   "vo-cutouts": 10
+               }
+           }
+       },
+       "bypass": [
+           "g_admins"
+       ]
+   }
+
+This mostly has the same structure as the configuration, but it overrides all quota information taken from the configuration, including additions from groups.
+So, for example, if the above override were in place, all users would have a quota of 10 for the datalinker API, including members of ``g_developers`` who normally get an extra 500.
+Members of the ``g_users`` group would only have a quota of 10 for the vo-cutouts API.
+
+There is an additional key here under notebook, ``spawn``, which is a boolean that controls whether affected users are allowed to spawn new notebooks at all.
+This allows quickly cutting off access to starting new Notebook Aspect pods for every user or only for users in particular groups, without changing token scopes.
+
+Finally, the ``bypass`` key in the ``quota-override`` data lists groups excluded from the override.
+In this example, members of ``g_admins`` can use the services according to the normal quota settings, without any changes from the override.
+
+In the initial implementation, Gafaelfawr won't cache the quota override information and will try to retrieve it from Redis for every request potentially affected by quotas.
+We'll see if that creates a performance problem and add in-memory caching if it does.
+
+API
+---
+
+There will be three new Gafaelfawr APIs to get and set the quota overrides:
+
+``GET /auth/api/v1/quota-overrides``
+    Retrieves the current quota overrides in the above JSON format.
+    Returns 404 if there are no quota overrides.
+
+``PUT /auth/api/v1/quota-overrides``
+    Creates or replaces the quota overrides.
+    The body should be the above JSON format.
+    There is no ``PATCH`` API; the complete override configuration has to be provided.
+    (We don't expect to need much complexity or to use this that frequently.)
+
+``DELETE /auth/api/v1/quota-overrides``
+    Delete the quota overrides.
+    Returns 404 if there are no quota overrides and 204 on success.
+
+Options considered
+------------------
+
+The original plan had been to store quota information in the database and provide an API and eventually a UI for updating it.
+However, that adds a lot of complexity and additional design for not a lot of expected benefit, at least for now.
```

Comment:
I appreciate the parallelism of "lot" and "not a lot" but in my head it sounds better with "significant" and "minimal" or something like that.